### PR TITLE
Handle non-existent files and custom file extensions

### PR DESCRIPTION
**Pull request by uniqueiniquity**
This PR addresses two main issues brought up in #50 when using the `--project` flag.
- If the file being parsed doesn't actually exist, the source file will not be found in the program generated from the tsconfig and the default, incomplete project will be returned.
- If the file being parsed has a non-traditional file extension (such as .vue), the source file will not be found in the program generated from the tsconfig and the default, incomplete project will be used.
While these two issues have the same general root cause (i.e. we build programs from the provided tsconfig(s) and then expect the file being parsed to be part of one of them), we address them separately.

For the former issue, we allow for a non-existent file to be used by parsing the compiler options out of a single tsconfig and building a program with that single, non-existent file and those compiler options. My intent was that this should be primarily used for test cases, which may provide code to be parsed instead of the contents of an actual file.

For the latter, we allow for a new string array option `extraFileExtensions`. This is then passed to the TypeScript API so that it will additionally include files with the provided file extensions when resolving the root files of a given program from a tsconfig. Unfortunately, this implementation uses an undocumented extension point in the TS API, so I will be making a change to TypeScript so this behavior is officially supported moving forward.

### CODE REVIEW

File: src/node-utils.ts [hunk 1]
```diff
@@ -205,14 +205,11 @@ function isESTreeClassMember(node: ts.Node): boolean {
...
- * @param {ts.KeywordSyntaxKind} modifierKind TypeScript SyntaxKind modifier
+ * @param {ts.SyntaxKind} modifierKind TypeScript SyntaxKind modifier
  * @param {ts.Node} node TypeScript AST node
  * @returns {boolean} has the modifier specified
  */
-function hasModifier(
-  modifierKind: ts.KeywordSyntaxKind,
-  node: ts.Node
-): boolean {
+function hasModifier(modifierKind: ts.SyntaxKind, node: ts.Node): boolean {
```
[the remark this answers is not in the document]
Comment (uniqueiniquity) [reply src/node-utils.ts]: Seems like I was working from an old version of the `typescript` type declarations that didn't have it, so I was getting errors. Changed it back, sorry about that.

**Comment by uniqueiniquity** [comment]
Also addresses concerns mentioned in https://github.com/eslint/typescript-eslint-parser/pull/568

**Comment by armano2** [comment]
@uniqueiniquity i tested it and seems to work, but can you add test scenario when you do not read file at all, but provide it as code with not existing filename?

**Comment by uniqueiniquity** [comment]
@armano2 thanks for taking a look. I intended `'non-existent file tests'` to test that scenario, since the filename I provided is `<input>`. Would you like me to add one with a not existing filename that looks more like a real filename?

**Comment by armano2** [comment]
.... i'm gettin blind i didn't see this test xd

**Pull request merged**